### PR TITLE
Support Scala3-NIGHTLY releases

### PR DESCRIPTION
**Pull request by dos65**
This introduces the following changes:
 - every release supports up to 5 of latest NIGHTLY Scala3 versions.
   Should be already valuable as every snapshot release will
   automatically catch new NIGHTLY releases.
 - `+cross/test` also run tests against the latest NIGHTLY release
  that should help with spotting issues earlier.

### CODE REVIEW

**Review by tgodzik — COMMENTED** [review]
I am a bit worried that this would pop up on the unrelated PRs.

Could we somehow only run it on master? Maybe by checking the current git branch ?

**Comment by dos65** [comment]
@tgodzik Yep, I was thinking about that. 
For the case if there is an issue in NIGHTLY release, my plan was to check the reason: compiler/metals. In case if it's compiler then blacklist this release somewhere in this function.

I think seeing failed tests for NIGHTLY only after merging in master might be more annoying than on PR stage.

Maybe let's try as it is? If it will be more problematic than useful we are always free to remove it from `+cross/test`.

File: build.sbt
Comment (dos65) [reply build.sbt:258]: Good idea, thx!

**Comment by ckipp01** [comment]
> I am a bit worried that this would pop up on the unrelated PRs.

>> I think seeing failed tests for NIGHTLY only after merging in master might be more annoying than on PR stage.

Yea I sort of agree with this, it might be hard to spot if it's only happening when it hits main, since most of. the time if there is a failed build on merge I don't go back really and look at what caused it. We could do something similar to our sourcegraph job where if it fails for a pr, then it doesn't block the merge or fail the check, but rather notify us that something was wrong.

>  In case if it's compiler then blacklist this release somewhere in this function.

Also, total aside, but let's try to replace our usage of _blacklist_ with just _deny_ since it is a loaded term.

File: build.sbt
Comment (tgodzik) [inline build.sbt:258]:
```suggestion
  lazy val nightlyScala3Versions =
```

File: build.sbt
Comment (tgodzik) [inline build.sbt:652]: Could we only run this on a separate job and remove the check from required for a merge? I will probably need to do it manually later on, but let's maybe have this run separately. This could be done via special env variable maybe?
```suggestion
    crossScalaVersions := {
         if (System.getenv("TEST_NIGHTLY") != null)  V.nightlyScala3Versions.lastOption.toList 
         else V.nonDeprecatedScalaVersions
     }
```
and add that variable to a separate workflow. What do you think?

**Review by tgodzik — CHANGES_REQUESTED** [review]

File: build.sbt
Comment (dos65) [reply build.sbt:652]: Good idea!
Will rework it a bit differently to call just a `++$LATEST_NIGHTLY cross/test` from a separate job

File: .github/workflows/ci.yml
Comment (dos65) [inline .github/workflows/ci.yml:5]: Some non-relevant changes were triggered by formatting in VScode but looks like they are valid.

File: build.sbt
Comment (tgodzik) [inline build.sbt:299]:
```suggestion
  lazy val nightlyScala3Versions = { if (isCi)
    Scala3NightlyVersions.nightlyReleasesAfter(nextScala3RC)
    else Nil
  }
```

To avoid fetching the versions when starting sbt locally, which would add to load time. I would even add a separate variable so that this is only added when needed in the nightlies cross tests. Otherwise, each job will try to download the versions.

So maybe:
```suggestion
  lazy val nightlyScala3Versions = { if (System.getenv("NIGHTLIES") != null)
    Scala3NightlyVersions.nightlyReleasesAfter(nextScala3RC)
    else Nil
  }
```
and set `NIGHTLIES` for the single cross test and for the releases.

File: build.sbt
Comment (tgodzik) [inline build.sbt:220]: Would we want to test it for all last 5 versions?

File: build.sbt
Comment (tgodzik) [inline build.sbt:184]: Wouldn't this be easier with the env var ?

File: build.sbt
Comment (tgodzik) [inline build.sbt:379]: With this, why would we need `configureMtagsScalaVersionDynamically` ?

**Review by tgodzik — CHANGES_REQUESTED** [review]

File: build.sbt
Comment (dos65) [reply build.sbt:184]: This function is anyway will be added in #3281 so I think it's better to use it as command.

File: build.sbt
Comment (dos65) [reply build.sbt:220]: I think having latest should be enough. Will save some minutes per ci run.

File: build.sbt
Comment (dos65) [reply build.sbt:299]: At the beginning I was thinking about additional costs from versions fetching but I was surprised by how fast it works an decided that probably it doesn't worth to prevent this call.

Just checked it more time: it takes less than 500ms.  Tested it with and without internet connection.
Maybe let's leave it as it is?

Comment (tgodzik) [reply build.sbt:299]: > it takes less than 500ms.

The problem is that at some point it all accumulates and sbt just loads super slow.
I would still change it to at least only run on CI, it's simple enough if anyone needs to make it work locally.

File: build.sbt
Comment (dos65) [reply build.sbt:379]: This is needed to publish mtags for nightly versions.
So at the release moment of for snapshot supported scala versions will be: all explicitly defined scala versions + up to 5 last nightly

**Review by kpodsiad — APPROVED** [review]
LGTM (includes Tomek's suggestions)

File: build.sbt
Comment (dos65) [reply build.sbt:299]: :smiling_face_with_tear: 
I understand your concerns but I think it will be confusing that you need to pass a special env to work with nightlies locally (test/debug/publish local). It also should be documented for contributors. 

Wdyt if I cover this thing by sbt cache? Probably it will be more faster.
The worst case will be that once a day coursier downloads this file(https://repo1.maven.org/maven2/org/scala-lang/scala3-compiler_3/maven-metadata.xml)   that I think should be about 500ms.

Comment (tgodzik) [reply build.sbt:299]: > I understand your concerns but I think it will be confusing that you need to pass a special env to work with nightlies locally (test/debug/publish local). It also should be documented for contributors.

To publish a nightly version all you need to do is `++3.0.1-RC1-bin-20210413-f3c1468-NIGHTLY` and `mtags/publishLocal`, so this wouldn't be a problem for releasing locally really. And with these changes when someone does `+publishLocal` they will have 5 more versions to publish each time even if everything was compiled last time.

I would really opt that it's all done in CI only.

Comment (dos65) [reply build.sbt]: Changed it - now it's needed  to set either `CI` or `NIGHTLIES` env var to enable this logic.

**Review by tgodzik — APPROVED** [review]
LGTM!

**Pull request merged**
